src/charm.py

```diff
@@ -1,104 +1,400 @@
+
+        Updates userlist.txt and pgbouncer.ini config files, reloading pgbouncer application once
+        updated.
+
+        Params:
+            users: a dictionary of usernames and passwords
+        """
+        if users is None:
+            users = self._get_userlist_from_container()
         else:
-            event.set_results({"fortune": "A bug in the code is worth two in the documentation."})
+            self._push_userlist(users)
+
+        self._push_pgbouncer_ini(users, reload_pgbouncer=True)
+
+    def _push_pgbouncer_ini(self, users=None, reload_pgbouncer=False) -> None:
+        """Generate pgbouncer.ini from juju config and deploy it to the container.
+
+        Params:
+            users: a dictionary of usernames and passwords
+            reload_pgbouncer: A boolean defining whether or not to reload the pgbouncer application
+                in the container. When config files are updated, pgbouncer must be restarted for
+                the changes to take effect. However, these config updates can be done in batches,
+                minimising the amount of necessary restarts.
+        """
+        if users is None:
+            users = self._get_userlist_from_container()
+
+        pgb_container = self.unit.get_container(self._pgbouncer_service)
+        pgbouncer_ini = self._generate_pgbouncer_ini(users)
+
+        try:
+            # Check that we're not updating this file unnecessarily
+            if pgb_container.pull(INI_PATH).read() == pgbouncer_ini:
+                logger.info("updated config does not modify existing pgbouncer config")
+                return
+        except FileNotFoundError:
+            # There is no existing pgbouncer.ini file, so carry on and add one.
+            pass
+
+        pgb_container.push(
+            INI_PATH,
+            pgbouncer_ini,
+            user=self._pgbouncer_user,
+            permissions=0o600,
+            make_dirs=True,
+        )
+        logging.info("pushed new pgbouncer.ini config file to pgbouncer container")
+
+        if reload_pgbouncer:
+            self._reload_pgbouncer()
+
+    def _generate_pgbouncer_ini(self, users=None) -> str:
+        """Generate pgbouncer.ini from config.
+
+        This is a basic stub method, and will be updated in future to generate more complex
+        pgbouncer.ini files in a more sophisticated way.
+
+        TODO compare to real-world pgbouncer.ini files and devise a more sophisticated way of
+        generating them. Look at the previous charm's implementation of this.
+        TODO evaluate other user types, such as stat_users etc
+        TODO define/parse databases based on relation to postgres - does it need to be an exposed
+        config option?
+
+        Params:
+            users: a dictionary of usernames and passwords
+        """
+        if users is None:
+            users = self._get_userlist_from_container()
+
+        return f"""[databases]
```

Comment:
This function is meant to implement a basic pgbouncer.ini that responds to basic juju config options so the pgbouncer service is capable of coming online. In future, a more sophisticated solution will be implemented that generates this file and handles its own database config from the postgres relation, but for now I'm comfortable using this as a placeholder, since it's nowhere near production-ready yet.

Reply:
Yeh I think ultimately the answer here will be to model the pgbouncer config as a class that is able to generate a valid config based on its attributes - but it's early days right now and there are few options supported so this is probably fine

Reply:
Sounds good to me!